scale/node/resources/node_resources.py

```diff
@@ -0,0 +1,210 @@
+        :type node_resources: :class:`node.resources.NodeResources`
+        """
+
+        for resource in node_resources.resources:
+            if resource.name in self._resources:
+                if self._resources[resource.name].value < resource.value:  # Assumes SCALAR type
+                    self._resources[resource.name].value = resource.value
+            else:
+                self._resources[resource.name] = resource.copy()
+
+    def is_equal(self, node_resources):
+        """Indicates if these resources are equal. This should be used for testing only.
+
+        :param node_resources: The resources to compare
+        :type node_resources: :class:`node.resources.NodeResources`
+        :returns: True if these resources are equal, False otherwise
+        :rtype: bool
+        """
+
+        # Make sure they have the exact same set of resource names
+        names = set()
+        for resource in node_resources.resources:
+            names.add(resource.name)
+        if set(self._resources.keys()) != names:
+            return False
+
+        for resource in node_resources.resources:
+            if round(self._resources[resource.name].value, 5) != round(resource.value, 5):  # Assumes SCALAR type
+                return False
+
+        return True
+
+    def is_sufficient_to_meet(self, node_resources):
+        """Indicates if these resources are sufficient to meet the requested resources
+
+        :param node_resources: The requested resources
+        :type node_resources: :class:`node.resources.NodeResources`
+        :returns: True if these resources are sufficient for the request, False otherwise
+        :rtype: bool
+        """
+
+        for resource in node_resources.resources:
+            if resource.name in self._resources:
+                if self._resources[resource.name].value < resource.value:  # Assumes SCALAR type
+                    return False
+            else:
+                # Do not have this resource, not a problem if requesting 0.0
+                if resource.value > 0.0:
+                    return False
+
+        return True
+
+    def remove_resource(self, name):
+        """Removes the resource with the given name
+
+        :param name: The name of the resource to remove
+        :type name: string
+        """
+
+        if name in self._resources:
+            if name in ['cpus', 'mem', 'disk']:
+                self._resources[name].value = 0.0
+            else:
+                del self._resources[name]
+
+    def subtract(self, node_resources):
+        """Subtracts the given resources
+
+        :param node_resources: The resources to subtract
+        :type node_resources: :class:`node.resources.NodeResources`
+        """
+
+        for resource in node_resources.resources:
+            if resource.name in self._resources:
+                self._resources[resource.name].value -= resource.value  # Assumes SCALAR type
+
+    def to_logging_string(self):
```

Comment:
You should probably just implement this as the `__str__` or `__repr__` magic methods. This will just allow you to pass it straight to logging with `%s` formatters.  I'd lean towards `__str__` for your case. Ref: https://stackoverflow.com/questions/1436703/difference-between-str-and-repr-in-python